### ISSUE TITLE
`make_adjoint` introducing `LocalLoad`s not taking `Alloca`s as inputs after implementing some optimization

### ISSUE THREAD

**Issue by xumingkuan** [taichi-dev/taichi#726]
After implementing the optimization of moving common statements in both branches outside `if`'s, I got something invalid like this:
```
<f32 x1> $57 = local load [ [$0[0]]]
<f32 x1> $58 = local load [ [$57[0]]]
```

(probably not) minimal test example:
```python
ti.init(print_preprocessed=True, print_ir=True)
x = ti.var(ti.f32, shape=2)
y = ti.var(ti.f32, shape=2)
ti.root.lazy_grad()
@ti.kernel
def func(i: ti.i32):
    t = x[i]
    if t > 0:
        y[i] = t
    else:
        y[i] = 2 * t
func.grad(0)
```

Debugging change set:
```C++
void make_adjoint(IRNode *root, bool use_stack) {
  if (use_stack) {
    fix_block_parents(root);
    ConvertLocalVar converter;
    root->accept(&converter);
    typecheck(root);

    std::cout << "before MakeAdjoint::run(root)\n";
    irpass::re_id(root);
    irpass::print(root);

    MakeAdjoint::run(root);

    std::cout << "after MakeAdjoint::run(root)\n";
    irpass::re_id(root);
    irpass::print(root);

    typecheck(root);
    fix_block_parents(root);

    std::cout << "after fix_block_parents(root)\n";
    irpass::re_id(root);
    irpass::print(root);

    BackupSSA b;
    root->accept(&b);
    typecheck(root);

    std::cout << "after make_adjoint\n";
    irpass::re_id(root);
    irpass::print(root);
  } else {
    MakeAdjoint::run(root);
    typecheck(root);
  }
}
```

Output:
```
before MakeAdjoint::run(root)
==========
kernel {
  <f32 x1> $0 = stack alloc (max_size=16)
  <i32 x1> $1 = arg[0]
  <f32*x1> $2 = global ptr [S2place_f32], index [$1] activate=true
  <f32 x1> $3 = global load $2
  <f32 x1> $4 = stack push $0, val = $3
  <i32 x1> $5 = const [1]
  <f32 x1> $6 = const [0.0]
  <i32 x1> $7 = cmp_gt $3 $6
  <i32 x1> $8 = bit_and $5 $7
  <i32 x1> $9 = alloca
  <i32 x1> $10 = alloca
  <f32 x1> $11 = stack load top $0
  if $8 {
    <i32 x1> $13 = arg[0]
    <f32*x1> $14 = global ptr [S4place_f32], index [$13] activate=true
    <f32*x1> $15 : global store [$14 <- $11]
  } else {
    <f32 x1> $16 = const [2.0]
    <f32 x1> $17 = mul $11 $16
    <i32 x1> $18 = arg[0]
    <f32*x1> $19 = global ptr [S4place_f32], index [$18] activate=true
    <f32*x1> $20 : global store [$19 <- $17]
  }
}
==========
after MakeAdjoint::run(root)
==========
kernel {
  <f32 x1> $0 = alloca
  <f32 x1> $1 = stack alloc (max_size=16)
  <i32 x1> $2 = arg[0]
  <f32*x1> $3 = global ptr [S2place_f32], index [$2] activate=true
  <f32 x1> $4 = global load $3
  <f32 x1> $5 = stack push $1, val = $4
  <i32 x1> $6 = const [1]
  <f32 x1> $7 = const [0.0]
  <i32 x1> $8 = cmp_gt $4 $7
  <i32 x1> $9 = bit_and $6 $8
  <i32 x1> $10 = alloca
  <i32 x1> $11 = alloca
  <f32 x1> $12 = stack load top $1
  if $9 {
    <i32 x1> $14 = arg[0]
    <f32*x1> $15 = global ptr [S4place_f32], index [$14] activate=true
  } else {
    <f32 x1> $16 = const [2.0]
    <f32 x1> $17 = mul $12 $16
    <i32 x1> $18 = arg[0]
    <f32*x1> $19 = global ptr [S4place_f32], index [$18] activate=true
  }
  if $9 {
    <f32 x1> $21 = alloca
    <f32 x1> $22 = global ptr [S6place_f32], index [$14] activate=true
    $23 = global load $22
    $24 = local load [ [$21[0]]]
    $25 = add $24 $23
    $26 : local store [$21 <- $25]
  } else {
    <f32 x1> $27 = alloca
    <f32 x1> $28 = alloca
    <f32 x1> $29 = global ptr [S6place_f32], index [$18] activate=true
    $30 = global load $29
    $31 = local load [ [$28[0]]]
    $32 = add $31 $30
    $33 : local store [$28 <- $32]
    $34 = local load [ [$28[0]]]
    $35 = mul $34 $16
    $36 = local load [ [$21[0]]]
    $37 = add $36 $35
    $38 : local store [$21 <- $37]
    $39 = local load [ [$28[0]]]
    $40 = mul $39 $12
    $41 = local load [ [$27[0]]]
    $42 = add $41 $40
    $43 : local store [$27 <- $42]
  }
  $44 = local load [ [$21[0]]]
  $45 : stack acc adj $1, val = $44
  $46 = stack load top adj $1
  $47 = local load [ [$0[0]]]
  $48 = add $47 $46
  $49 : local store [$0 <- $48]
  $50 : stack pop $1
  <f32 x1> $51 = global ptr [S5place_f32], index [$2] activate=true
  $52 = local load [ [$0[0]]]
  $53 = atomic add($51, $52)
}
==========
after fix_block_parents(root)
==========
kernel {
  <f32 x1> $0 = alloca
  <f32 x1> $1 = stack alloc (max_size=16)
  <i32 x1> $2 = arg[0]
  <f32*x1> $3 = global ptr [S2place_f32], index [$2] activate=true
  <f32 x1> $4 = global load $3
  <f32 x1> $5 = stack push $1, val = $4
  <i32 x1> $6 = const [1]
  <f32 x1> $7 = const [0.0]
  <i32 x1> $8 = cmp_gt $4 $7
  <i32 x1> $9 = bit_and $6 $8
  <i32 x1> $10 = alloca
  <i32 x1> $11 = alloca
  <f32 x1> $12 = stack load top $1
  if $9 {
    <i32 x1> $14 = arg[0]
    <f32*x1> $15 = global ptr [S4place_f32], index [$14] activate=true
  } else {
    <f32 x1> $16 = const [2.0]
    <f32 x1> $17 = mul $12 $16
    <i32 x1> $18 = arg[0]
    <f32*x1> $19 = global ptr [S4place_f32], index [$18] activate=true
  }
  if $9 {
    <f32 x1> $21 = alloca
    <f32*x1> $22 = global ptr [S6place_f32], index [$14] activate=true
    <f32 x1> $23 = global load $22
    <f32 x1> $24 = local load [ [$21[0]]]
    <f32 x1> $25 = add $24 $23
    <f32 x1> $26 : local store [$21 <- $25]
  } else {
    <f32 x1> $27 = alloca
    <f32 x1> $28 = alloca
    <f32*x1> $29 = global ptr [S6place_f32], index [$18] activate=true
    <f32 x1> $30 = global load $29
    <f32 x1> $31 = local load [ [$28[0]]]
    <f32 x1> $32 = add $31 $30
    <f32 x1> $33 : local store [$28 <- $32]
    <f32 x1> $34 = local load [ [$28[0]]]
    <f32 x1> $35 = mul $34 $16
    <f32 x1> $36 = local load [ [$21[0]]]
    <f32 x1> $37 = add $36 $35
    <f32 x1> $38 : local store [$21 <- $37]
    <f32 x1> $39 = local load [ [$28[0]]]
    <f32 x1> $40 = mul $39 $12
    <f32 x1> $41 = local load [ [$27[0]]]
    <f32 x1> $42 = add $41 $40
    <f32 x1> $43 : local store [$27 <- $42]
  }
  <f32 x1> $44 = local load [ [$21[0]]]
  <f32 x1> $45 : stack acc adj $1, val = $44
  <f32 x1> $46 = stack load top adj $1
  <f32 x1> $47 = local load [ [$0[0]]]
  <f32 x1> $48 = add $47 $46
  <f32 x1> $49 : local store [$0 <- $48]
  <f32 x1> $50 : stack pop $1
  <f32*x1> $51 = global ptr [S5place_f32], index [$2] activate=true
  <f32 x1> $52 = local load [ [$0[0]]]
  <f32 x1> $53 = atomic add($51, $52)
}
==========
after make_adjoint
==========
kernel {
  <f32 x1> $0 = alloca
  <f32 x1> $1 = alloca
  <i32 x1> $2 = alloca
  <i32 x1> $3 = alloca
  <f32 x1> $4 = alloca
  <f32 x1> $5 = stack alloc (max_size=16)
  <i32 x1> $6 = arg[0]
  <f32*x1> $7 = global ptr [S2place_f32], index [$6] activate=true
  <f32 x1> $8 = global load $7
  <f32 x1> $9 = stack push $5, val = $8
  <i32 x1> $10 = const [1]
  <f32 x1> $11 = const [0.0]
  <i32 x1> $12 = cmp_gt $8 $11
  <i32 x1> $13 = bit_and $10 $12
  <i32 x1> $14 = alloca
  <i32 x1> $15 = alloca
  <f32 x1> $16 = stack load top $5
  if $13 {
    <i32 x1> $18 = arg[0]
    <i32 x1> $19 : local store [$3 <- $18]
    <f32*x1> $20 = global ptr [S4place_f32], index [$18] activate=true
  } else {
    <f32 x1> $21 = const [2.0]
    <f32 x1> $22 : local store [$1 <- $21]
    <f32 x1> $23 = mul $16 $21
    <i32 x1> $24 = arg[0]
    <i32 x1> $25 : local store [$2 <- $24]
    <f32*x1> $26 = global ptr [S4place_f32], index [$24] activate=true
  }
  if $13 {
    <f32 x1> $28 = alloca
    <f32 x1> $29 : local store [$0 <- $28]
    <i32 x1> $30 = local load [ [$3[0]]]
    <f32*x1> $31 = global ptr [S6place_f32], index [$30] activate=true
    <f32 x1> $32 = global load $31
    <f32 x1> $33 = local load [ [$28[0]]]
    <f32 x1> $34 = add $33 $32
    <f32 x1> $35 : local store [$28 <- $34]
  } else {
    <f32 x1> $36 = alloca
    <f32 x1> $37 = alloca
    <i32 x1> $38 = local load [ [$2[0]]]
    <f32*x1> $39 = global ptr [S6place_f32], index [$38] activate=true
    <f32 x1> $40 = global load $39
    <f32 x1> $41 = local load [ [$37[0]]]
    <f32 x1> $42 = add $41 $40
    <f32 x1> $43 : local store [$37 <- $42]
    <f32 x1> $44 = local load [ [$37[0]]]
    <f32 x1> $45 = local load [ [$1[0]]]
    <f32 x1> $46 = mul $44 $45
    <f32 x1> $47 = local load [ [$0[0]]]
    <f32 x1> $48 = local load [ [$47[0]]]
    <f32 x1> $49 = add $48 $46
    <f32 x1> $50 = local load [ [$0[0]]]
    <f32 x1> $51 : local store [$50 <- $49]
    <f32 x1> $52 = local load [ [$37[0]]]
    <f32 x1> $53 = mul $52 $16
    <f32 x1> $54 = local load [ [$36[0]]]
    <f32 x1> $55 = add $54 $53
    <f32 x1> $56 : local store [$36 <- $55]
  }
  <f32 x1> $57 = local load [ [$0[0]]]
  <f32 x1> $58 = local load [ [$57[0]]]
  <f32 x1> $59 : stack acc adj $5, val = $58
  <f32 x1> $60 = stack load top adj $5
  <f32 x1> $61 = local load [ [$4[0]]]
  <f32 x1> $62 = add $61 $60
  <f32 x1> $63 : local store [$4 <- $62]
  <f32 x1> $64 : stack pop $5
  <f32*x1> $65 = global ptr [S5place_f32], index [$6] activate=true
  <f32 x1> $66 = local load [ [$4[0]]]
  <f32 x1> $67 = atomic add($65, $66)
}
==========
```
The IR before `MakeAdjoint::run(root)` looks good to me.

The IR after `MakeAdjoint::run(root)` looks weird: `$36 = local load [ [$21[0]]]` seems invalid.

And the IR after the whole function is invalid: it contains a local load of a local load.

I'm not familiar with `MakeAdjoint`. @yuanming-hu do you have any ideas?

**Comment by yuanming-hu** [comment]
I'm not 100% sure if it is caused by something inside `MakeAdjoint`. One possibility is that something outside that pass caused an undefined behavior. Could you open a Draft PR for your optimization implementation?

**Comment by yuanming-hu** [comment]
The updated #734 with #727 now gives

```
[I 04/09/20 18:26:10.389] [compile_to_offloads.cpp:operator()@17] Adjoint 1:
==========
kernel {
  <f32 x1> $0 = alloca
  <i32 x1> $1 = arg[0]
  <f32*x1> $2 = global ptr [S2place_f32], index [$1] activate=true
  <f32 x1> $3 = global load $2
  <f32 x1> $4 : local store [$0 <- $3]
  <i32 x1> $5 = const [1]
  <f32 x1> $6 = const [0.0]
  <i32 x1> $7 = cmp_gt $3 $6
  <i32 x1> $8 = bit_and $5 $7
  <i32 x1> $9 = alloca
  <i32 x1> $10 = alloca
  <f32 x1> $11 = local load [ [$0[0]]]
  if $8 {
    <i32 x1> $13 = arg[0]
    <f32*x1> $14 = global ptr [S4place_f32], index [$13] activate=true
    <f32*x1> $15 : global store [$14 <- $11]
  } else {
    <f32 x1> $16 = const [2.0]
    <f32 x1> $17 = mul $11 $16
    <i32 x1> $18 = arg[0]
    <f32*x1> $19 = global ptr [S4place_f32], index [$18] activate=true
    <f32*x1> $20 : global store [$19 <- $17]
  }
}
==========
before MakeAdjoint::run(root)
==========
kernel {
  <f32 x1> $0 = stack alloc (max_size=16)
  <i32 x1> $1 = arg[0]
  <f32*x1> $2 = global ptr [S2place_f32], index [$1] activate=true
  <f32 x1> $3 = global load $2
  <f32 x1> $4 = stack push $0, val = $3
  <i32 x1> $5 = const [1]
  <f32 x1> $6 = const [0.0]
  <i32 x1> $7 = cmp_gt $3 $6
  <i32 x1> $8 = bit_and $5 $7
  <i32 x1> $9 = alloca
  <i32 x1> $10 = alloca
  <f32 x1> $11 = stack load top $0
  if $8 {
    <i32 x1> $13 = arg[0]
    <f32*x1> $14 = global ptr [S4place_f32], index [$13] activate=true
    <f32*x1> $15 : global store [$14 <- $11]
  } else {
    <f32 x1> $16 = const [2.0]
    <f32 x1> $17 = mul $11 $16
    <i32 x1> $18 = arg[0]
    <f32*x1> $19 = global ptr [S4place_f32], index [$18] activate=true
    <f32*x1> $20 : global store [$19 <- $17]
  }
}
==========
after MakeAdjoint::run(root)
==========
kernel {
  <f32 x1> $0 = alloca
  <f32 x1> $1 = alloca
  <f32 x1> $2 = alloca
  <f32 x1> $3 = alloca
  <f32 x1> $4 = stack alloc (max_size=16)
  <i32 x1> $5 = arg[0]
  <f32*x1> $6 = global ptr [S2place_f32], index [$5] activate=true
  <f32 x1> $7 = global load $6
  <f32 x1> $8 = stack push $4, val = $7
  <i32 x1> $9 = const [1]
  <f32 x1> $10 = const [0.0]
  <i32 x1> $11 = cmp_gt $7 $10
  <i32 x1> $12 = bit_and $9 $11
  <i32 x1> $13 = alloca
  <i32 x1> $14 = alloca
  <f32 x1> $15 = stack load top $4
  if $12 {
    <i32 x1> $17 = arg[0]
    <f32*x1> $18 = global ptr [S4place_f32], index [$17] activate=true
  } else {
    <f32 x1> $19 = const [2.0]
    <f32 x1> $20 = mul $15 $19
    <i32 x1> $21 = arg[0]
    <f32*x1> $22 = global ptr [S4place_f32], index [$21] activate=true
  }
  if $12 {
    <f32 x1> $24 = global ptr [S6place_f32], index [$17] activate=true
    $25 = global load $24
    $26 = local load [ [$3[0]]]
    $27 = add $26 $25
    $28 : local store [$3 <- $27]
  } else {
    <f32 x1> $29 = global ptr [S6place_f32], index [$21] activate=true
    $30 = global load $29
    $31 = local load [ [$2[0]]]
    $32 = add $31 $30
    $33 : local store [$2 <- $32]
    $34 = local load [ [$2[0]]]
    $35 = mul $34 $19
    $36 = local load [ [$3[0]]]
    $37 = add $36 $35
    $38 : local store [$3 <- $37]
    $39 = local load [ [$2[0]]]
    $40 = mul $39 $15
    $41 = local load [ [$1[0]]]
    $42 = add $41 $40
    $43 : local store [$1 <- $42]
  }
  $44 = local load [ [$3[0]]]
  $45 : stack acc adj $4, val = $44
  $46 = stack load top adj $4
  $47 = local load [ [$0[0]]]
  $48 = add $47 $46
  $49 : local store [$0 <- $48]
  $50 : stack pop $4
  <f32 x1> $51 = global ptr [S5place_f32], index [$5] activate=true
  $52 = local load [ [$0[0]]]
  $53 = atomic add($51, $52)
}
==========
after fix_block_parents(root)
==========
kernel {
  <f32 x1> $0 = alloca
  <f32 x1> $1 = alloca
  <f32 x1> $2 = alloca
  <f32 x1> $3 = alloca
  <f32 x1> $4 = stack alloc (max_size=16)
  <i32 x1> $5 = arg[0]
  <f32*x1> $6 = global ptr [S2place_f32], index [$5] activate=true
  <f32 x1> $7 = global load $6
  <f32 x1> $8 = stack push $4, val = $7
  <i32 x1> $9 = const [1]
  <f32 x1> $10 = const [0.0]
  <i32 x1> $11 = cmp_gt $7 $10
  <i32 x1> $12 = bit_and $9 $11
  <i32 x1> $13 = alloca
  <i32 x1> $14 = alloca
  <f32 x1> $15 = stack load top $4
  if $12 {
    <i32 x1> $17 = arg[0]
    <f32*x1> $18 = global ptr [S4place_f32], index [$17] activate=true
  } else {
    <f32 x1> $19 = const [2.0]
    <f32 x1> $20 = mul $15 $19
    <i32 x1> $21 = arg[0]
    <f32*x1> $22 = global ptr [S4place_f32], index [$21] activate=true
  }
  if $12 {
    <f32*x1> $24 = global ptr [S6place_f32], index [$17] activate=true
    <f32 x1> $25 = global load $24
    <f32 x1> $26 = local load [ [$3[0]]]
    <f32 x1> $27 = add $26 $25
    <f32 x1> $28 : local store [$3 <- $27]
  } else {
    <f32*x1> $29 = global ptr [S6place_f32], index [$21] activate=true
    <f32 x1> $30 = global load $29
    <f32 x1> $31 = local load [ [$2[0]]]
    <f32 x1> $32 = add $31 $30
    <f32 x1> $33 : local store [$2 <- $32]
    <f32 x1> $34 = local load [ [$2[0]]]
    <f32 x1> $35 = mul $34 $19
    <f32 x1> $36 = local load [ [$3[0]]]
    <f32 x1> $37 = add $36 $35
    <f32 x1> $38 : local store [$3 <- $37]
    <f32 x1> $39 = local load [ [$2[0]]]
    <f32 x1> $40 = mul $39 $15
    <f32 x1> $41 = local load [ [$1[0]]]
    <f32 x1> $42 = add $41 $40
    <f32 x1> $43 : local store [$1 <- $42]
  }
  <f32 x1> $44 = local load [ [$3[0]]]
  <f32 x1> $45 : stack acc adj $4, val = $44
  <f32 x1> $46 = stack load top adj $4
  <f32 x1> $47 = local load [ [$0[0]]]
  <f32 x1> $48 = add $47 $46
  <f32 x1> $49 : local store [$0 <- $48]
  <f32 x1> $50 : stack pop $4
  <f32*x1> $51 = global ptr [S5place_f32], index [$5] activate=true
  <f32 x1> $52 = local load [ [$0[0]]]
  <f32 x1> $53 = atomic add($51, $52)
}
==========
after make_adjoint
==========
kernel {
  <f32 x1> $0 = alloca
  <i32 x1> $1 = alloca
  <i32 x1> $2 = alloca
  <f32 x1> $3 = alloca
  <f32 x1> $4 = alloca
  <f32 x1> $5 = alloca
  <f32 x1> $6 = alloca
  <f32 x1> $7 = stack alloc (max_size=16)
  <i32 x1> $8 = arg[0]
  <f32*x1> $9 = global ptr [S2place_f32], index [$8] activate=true
  <f32 x1> $10 = global load $9
  <f32 x1> $11 = stack push $7, val = $10
  <i32 x1> $12 = const [1]
  <f32 x1> $13 = const [0.0]
  <i32 x1> $14 = cmp_gt $10 $13
  <i32 x1> $15 = bit_and $12 $14
  <i32 x1> $16 = alloca
  <i32 x1> $17 = alloca
  <f32 x1> $18 = stack load top $7
  if $15 {
    <i32 x1> $20 = arg[0]
    <i32 x1> $21 : local store [$2 <- $20]
    <f32*x1> $22 = global ptr [S4place_f32], index [$20] activate=true
  } else {
    <f32 x1> $23 = const [2.0]
    <f32 x1> $24 : local store [$0 <- $23]
    <f32 x1> $25 = mul $18 $23
    <i32 x1> $26 = arg[0]
    <i32 x1> $27 : local store [$1 <- $26]
    <f32*x1> $28 = global ptr [S4place_f32], index [$26] activate=true
  }
  if $15 {
    <i32 x1> $30 = local load [ [$2[0]]]
    <f32*x1> $31 = global ptr [S6place_f32], index [$30] activate=true
    <f32 x1> $32 = global load $31
    <f32 x1> $33 = local load [ [$6[0]]]
    <f32 x1> $34 = add $33 $32
    <f32 x1> $35 : local store [$6 <- $34]
  } else {
    <i32 x1> $36 = local load [ [$1[0]]]
    <f32*x1> $37 = global ptr [S6place_f32], index [$36] activate=true
    <f32 x1> $38 = global load $37
    <f32 x1> $39 = local load [ [$5[0]]]
    <f32 x1> $40 = add $39 $38
    <f32 x1> $41 : local store [$5 <- $40]
    <f32 x1> $42 = local load [ [$5[0]]]
    <f32 x1> $43 = local load [ [$0[0]]]
    <f32 x1> $44 = mul $42 $43
    <f32 x1> $45 = local load [ [$6[0]]]
    <f32 x1> $46 = add $45 $44
    <f32 x1> $47 : local store [$6 <- $46]
    <f32 x1> $48 = local load [ [$5[0]]]
    <f32 x1> $49 = mul $48 $18
    <f32 x1> $50 = local load [ [$4[0]]]
    <f32 x1> $51 = add $50 $49
    <f32 x1> $52 : local store [$4 <- $51]
  }
  <f32 x1> $53 = local load [ [$6[0]]]
  <f32 x1> $54 : stack acc adj $7, val = $53
  <f32 x1> $55 = stack load top adj $7
  <f32 x1> $56 = local load [ [$3[0]]]
  <f32 x1> $57 = add $56 $55
  <f32 x1> $58 : local store [$3 <- $57]
  <f32 x1> $59 : stack pop $7
  <f32*x1> $60 = global ptr [S5place_f32], index [$8] activate=true
  <f32 x1> $61 = local load [ [$3[0]]]
  <f32 x1> $62 = atomic add($60, $61)
}
==========

```

It seems to me that the invalid statements are gone, but my brian is half-fried so my judgment probably doesn't count. We should use your verification pass to check...

**Comment by xumingkuan** [comment]
Thanks! It works now. Just wait and see if #727 passes CI.

**Comment by xumingkuan** [comment]
Closing as fixed in #734 with #727.
